Define callbacks on descendants.

rywall opened the pull request:
### Summary

We call `set_callbacks` on all descendants, so we need to make sure that callbacks are also defined on all descendants. Otherwise, you will get an error when running `set_callback` for a callback that was defined after subclasses were added.

```ruby
class Foo < ActiveRecord::Base
end

ActiveRecord::Base.define_model_callbacks :greet
ActiveRecord::Base.after_greet { puts 'goodbye' }
=> NoMethodError: undefined method `prepend' for nil:NilClass
```

rywall committed:
Define callbacks on descendants.

We set callbacks on all descendants, so we need to make sure that they are also defined on all descendants as well.

njakobsen commented on activesupport/lib/active_support/callbacks.rb:
```diff
@@ -806,28 +806,30 @@ def reset_callbacks(name)
 
-          names.each do |name|
-            name = name.to_sym
+          ([self] + ActiveSupport::DescendantsTracker.descendants(self)).each do |target|
```
I wonder if this pattern is worth making a `subtree` method in `ActiveSupport::DescendantsTracker` that includes `self` and `descendants`.

rafaelfranca replied:
Not really, but it make sense to extract this method on this module since we use this same code in more than one place.

rywall replied:
@rafaelfranca Sure I could do that. Should I do it as part of this PR? Any other feedback about the PR?

rywall commented:
@pixeltrix @rafaelfranca Let me know if I can improve this PR in any way or explain the reasoning behind it a little more clearly. FYI, I tried to provide a concise example of the problem in my initial PR comment.

matthewd commented:
I don't think there's any need to define the methods on the descendants?

rywall commented:
@matthewd Is my test case an invalid use case?

```ruby
class Foo < ActiveRecord::Base
end

ActiveRecord::Base.define_model_callbacks :greet
ActiveRecord::Base.after_greet { puts 'goodbye' }
=> NoMethodError: undefined method `prepend' for nil:NilClass
```

rywall commented:
@matthewd Also, I tested that the test I wrote fails on the master branch.

rywall commented:
@matthewd @pixeltrix Also confirmed this is a regression in 5.1.

5-0-stable:
![image](https://user-images.githubusercontent.com/7904/36322512-7a22cddc-131b-11e8-921c-693e5cdeae11.png)

matthewd commented:
I understand the extra `set_callbacks` calls; I'm questioning the need for the method definitions.

rywall committed:
No need to define methods on descendants.

Addresses feedback from https://github.com/rails/rails/pull/31913#issuecomment-365983580

rywall commented:
@matthewd Ahh I understand now. Good point. I've updated the PR. What do you think now?

rywall commented:
@matthewd @pixeltrix Is there anything else I can do to address your questions/concerns about this PR?

rywall commented:
@matthewd @pixeltrix Please let me know if I can help get this PR merged. I could just submit the failing test case if you'd rather fix it in a different way?

rywall commented:
@matthewd @pixeltrix Hi just checking in on this PR again. Hoping to make a contribution. Would be willing to make any necessary changes.

matthewd commented:
Sorry this sat for a while. Looks great, thanks!

rywall commented:
@matthewd Thank you! 💛
